File: julia/src/Heuristics.jl
module Heuristics

export WarmStart, build_warm_start

using ..IOUtil: FlowSpec, ProblemSpec, ScheduleSegment
using ..Candidates: CandidateIndex, FlowCandidate, travel_time
using ..Capacity: capacity_lookup

const EPS = 1e-6

struct WarmStart
    assignments::Dict{Int, Vector{ScheduleSegment}}
end

function _score_cell(flow::FlowSpec, times::Vector{Int}, coord::Tuple{Int, Int}, capacity_table)
    total = 0.0
    for t in times
        total += capacity_lookup(capacity_table, coord, t)
    end
    return total / (1 + travel_time(flow, coord))
end

function _order_cells(flow::FlowSpec, fc::FlowCandidate, capacity_table)
    scored = [
        (coord, _score_cell(flow, times, coord, capacity_table))
        for (coord, times) in fc.slots
    ]
    sort!(scored; by = x -> x[2], rev = true)
    return [item[1] for item in scored]
end

function build_warm_start(problem::ProblemSpec, candidates::CandidateIndex, capacity_table)
    remaining = Dict{Tuple{Int, Int}, Vector{Float64}}()
    for (coord, caps) in capacity_table
        remaining[coord] = copy(caps)
    end

    assignments = Dict{Int, Vector{ScheduleSegment}}()
    flows_sorted = sort(problem.flows; by = f -> f.start_time)

    for flow in flows_sorted
        fc = candidates.flows[flow.id]
        segments = ScheduleSegment[]
        assigned = 0.0
        occupied_times = Set{Int}()

        ordered_cells = _order_cells(flow, fc, remaining)
        for coord in ordered_cells
            times = sort(fc.slots[coord])
            for t in times
                if t in occupied_times
                    continue
                end
                caps = remaining[coord]
                avail = caps[t + 1]
                if avail <= EPS
                    continue
                end
                needed = flow.volume - assigned
                if needed <= EPS
                    break
                end
                take = min(needed, avail)
                if take <= EPS
                    continue
                end
                caps[t + 1] -= take
                push!(segments, ScheduleSegment(t, coord[1], coord[2], take))
                push!(occupied_times, t)
                assigned += take
            end
            if assigned + EPS >= flow.volume
                break
            end
        end

        assignments[flow.id] = segments
    end

    return WarmStart(assignments)
end

end # module

File: src/heuristics.py
"""Greedy warm-start schedule respecting landing-zone rules."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Dict, List, Tuple

from .capacity import capacity_lookup
from .candidates import CandidateIndex, FlowCandidate
from .models import Flow, Problem, ScheduleSegment

EPSILON = 1e-6


@dataclass(slots=True)
class WarmStart:
    assignments: Dict[int, List[ScheduleSegment]]


def _score_landing_cell(flow: Flow, slots: List[int], cell: Tuple[int, int], capacity_table) -> float:
    total_capacity = sum(capacity_lookup(capacity_table, cell, time_slot) for time_slot in slots)
    travel = max(1, flow.travel_time(cell))
    return total_capacity / travel


def _ordered_cells(flow: Flow, candidate: FlowCandidate, capacity_table) -> List[Tuple[int, int]]:
    scored = [
        (cell, _score_landing_cell(flow, slots, cell, capacity_table))
        for cell, slots in candidate.time_slots.items()
    ]
    scored.sort(key=lambda item: item[1], reverse=True)
    return [cell for cell, _ in scored]


def build_warm_start(problem: Problem, index: CandidateIndex, capacity_table) -> WarmStart:
    remaining_capacity: Dict[Tuple[int, int], List[float]] = {}
    for cell, capacities in capacity_table.items():
        remaining_capacity[cell] = list(capacities)

    assignments: Dict[int, List[ScheduleSegment]] = {}
    flows_sorted = sorted(problem.flows, key=lambda flow: flow.start_time)

    for flow in flows_sorted:
        candidate = index.flows.get(flow.flow_id)
        if candidate is None:
            assignments[flow.flow_id] = []
            continue

        segments: List[ScheduleSegment] = []
        allocated = 0.0
        used_times: set[int] = set()

        for cell in _ordered_cells(flow, candidate, remaining_capacity):
            for time_slot in candidate.time_slots[cell]:
                if time_slot in used_times:
                    continue
                available = remaining_capacity[cell][time_slot]
                if available <= EPSILON:
                    continue
                remaining = flow.volume - allocated
                if remaining <= EPSILON:
                    break
                take = min(available, remaining)
                remaining_capacity[cell][time_slot] -= take
                segments.append(ScheduleSegment(time=time_slot, x=cell[0], y=cell[1], rate=take))
                used_times.add(time_slot)
                allocated += take
            if allocated + EPSILON >= flow.volume:
                break

        assignments[flow.flow_id] = segments

    return WarmStart(assignments=assignments)
